### PR TITLE
feature: s3 template merge

### PR DESCRIPTION
**Pull request by Benbentwo**
## what
* enhancement to s3-mirroring: template merging.

## why
* The infrastructure repository should handle the things like volume mounts family etc. the Application side should handle the image updates and any env variable changes. This change helps support that.

## references
* https://github.com/cloudposse/terraform-aws-components/pull/997

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -63,6 +65,52 @@ runs:
...
+    - name: Merge Task Definitions
+      if: ${{ inputs.operation == 'deploy' && inputs.mirror_to_s3_bucket != '' && inputs.template_from_s3_bucket != '' }}
+      shell: bash
+      run: |
+        TEMP="./${{ steps.tmp.outputs.dir_name }}"
+        jq -s '.[].containerDefinitions | map( { (.name|tostring): . } ) | add' $TEMP/s3_task_def.json > $TEMP/s3_container_map.json
+        jq -s '.[].containerDefinitions | map( { (.name|tostring): . } ) | add' ${{ inputs.taskdef-path }} > $TEMP/app_container_map.json
+        jq -s '.[0] * .[1]' $TEMP/s3_container_map.json $TEMP/app_container_map.json > $TEMP/mergedContainerDefinitionMap.json
+        jq -s '[.[] | to_entries] | flatten | [.[] | .value] | {"containerDefinitions": .}' $TEMP/mergedContainerDefinitionMap.json > $TEMP/containerDefinitions.json
+        jq -s '.[0] * .[1] * .[2]' $TEMP/s3_task_def.json ${{ inputs.taskdef-path }} $TEMP/containerDefinitions.json > ./${{ steps.tmp.outputs.dir_name }}/${{ inputs.application }}.json
```
Comment (osterman) [inline action.yml:101]: The following seems to work for me.

```shell
jq --slurp '
  {
    containerDefinitions: (
      reduce (.[0].containerDefinitions[] | {(.name): .}) as $item ({}; . + $item)
      *
      reduce (.[1].containerDefinitions[] | {(.name): .}) as $item ({}; . + $item)
    )
    | to_entries
    | map(.value)
  }
  + (.[0] + .[1] | del(.containerDefinitions))
' t2.json t1.json
```

File: action.yml [hunk 2]
```diff
@@ -63,6 +65,52 @@ runs:
...
+      shell: bash
+      run: |
+        VALUE=$(echo $RANDOM | md5sum | head -c 20)
+        echo "dir_name=${VALUE}" >> $GITHUB_OUTPUT
+
+    - name: Create TMP dir
+      if: ${{ inputs.operation == 'deploy' }}
+      shell: bash
+      run: |
+        mkdir ./${{ steps.tmp.outputs.dir_name }}
```
Comment (osterman) [inline action.yml:80]:
```suggestion
        mkdir -p ./${{ steps.tmp.outputs.dir_name }}
```

File: action.yml [hunk 3]
```diff
@@ -63,6 +65,52 @@ runs:
...
+    - name: Debug
+      if: ${{ inputs.operation == 'deploy' && inputs.mirror_to_s3_bucket != '' && inputs.template_from_s3_bucket != '' && inputs.debug == 'true'}}
+      shell: bash
+      run: |
+        echo "----------------- S3 Task Definition -----------------"
+        cat ./${{ steps.tmp.outputs.dir_name }}/s3_task_def.json | jq
+        echo "----------------- Current Task Definition -----------------"
+        cat ${{ inputs.taskdef-path }}
+        echo "----------------- Merged Task Definition -----------------"
+        cat ./${{ steps.tmp.outputs.dir_name }}/${{ inputs.application }}.json
```
Comment (osterman) [inline action.yml:112]: Consider using Step Summary instead.

File: README.yaml [hunk 4]
```diff
@@ -73,7 +73,18 @@ usage: |-
         outputs:
           result: ${{ steps.example.outputs.webapp-url }}
   ```
+  
+  ## S3 Mirroring
+  
+  This action supports s3 mirroring. This means the [`ecs-service`](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/ecs-service) component will upload a task definition template to S3, this action will use and merge that template with the task definition from the repository and then deploy the merged task definition to ECS.
```
Comment (osterman) [inline README.yaml:79]:
```suggestion
  This action supports synchronizing `taskDefinitions` to S3. This is designed to work in concert with Cloud Posse's [`ecs-service`](https://github.com/cloudposse/terraform-aws-components/tree/main/modules/ecs-service) component, that will upload a task definition template to S3. This action will then pull that template from S3 and merge it with the local task definition from the repository. Finally, `ecspresso` will be called to deploy the merged task definition to ECS.
```

**Review by goruha — APPROVED** [review]

**Pull request merged**